### DISCUSSION TITLE
...

### DISCUSSION THREAD

**Discussion by ghost** [vercel/next.js#52697] [Help] [answered]

**Comment by icyJoseph** [comment]
The problem is in the last lines, you need to do this:

```js
module.exports = (phase, { defaultConfig }) => {
  return withBundleAnalyzer(defaultConfig);
};
```

The issue is that the second argument to the next.config.js in function form, is an object, with a key named `defaultConfig`.

This info has gotten lost in the docs, but this old translations have it: https://nextjs-ja-translation-docs.vercel.app/docs/api-reference/next.config.js/introduction (just look at the code snippets)

**Reply by icyJoseph** [reply, marked as answer]
Right, typically the way to do this **_is_** to just compose them:

```ts
module.exports = withPluginA(withPluginB(withPluginC(myNextConfig)));
```

That, for some, already looks a bit weird, and some plugins added their own keys and options parameters, so you'd end up with something like

```ts
module.exports = withPluginA(withPluginB(myConfig), { /* some options for plugin A */ });
```

Which looks even worse now. And again, plugins might have used their own keys, so we sort of hoped that, if Plugin B had some special keys, then Plugin A, better pass those along too! 

> https://www.npmjs.com/package/next-compose-plugins a solution for the early 2020's

However, even if you used a composer, the more plugins you had, the uglier things looked.

To standardise this, some months ago Next.js create some warnings, to nudge plugin writers into following a standard object and call signature.

This way, plugins would first expose a factory function, which would take in the plugin specific config, and then we could just, go back to simply compose them:

```ts
const withPluginA = require("plugin-a")({/* options for plugin a */})
// or
const createPluginA = require("plugin-a")
const withPluginA = createPluginA({/* options for plugin a */})
```

So for example to compose, the bundle analyzer and the reveal plugin you'd do:

```js
// next.config.js
const createNextPluginPreval = require("next-plugin-preval/config");
const withNextPluginPreval = createNextPluginPreval();

const withBundleAnalyzer = require("@next/bundle-analyzer")({
  enabled: process.env.ANALYZE === "true",
});

const nextConfig = {
  reactStrictMode: true,
};

module.exports = withBundleAnalyzer(withNextPluginPreval(nextConfig));
```

And then with some preval file like:

```js
import preval from "next-plugin-preval";

async function getData() {
  return { hello: "world" };
}

export default preval(await getData());
```

Then running `ANALYZE=true yarn build && yarn start` prints the bundle mapping, and uses the preval data hello world 😄 - that command works only on windows because of the `ANALYZE=true` inlined at the beginning.

**Reply by icyJoseph** [reply]
@terlan4 please don't forget to also mark as answered 🙏
